### ISSUE TITLE
Enable optimisations with `Chain`

### ISSUE THREAD

**Issue by theabhirath** [FluxML/Flux.jl#2004]
In long models in Metalhead, it is often the case that there are layers that can be reduced to `identity` - `Dropout(p = 0)` is a frequent occurrence, along with some other similar regularisation layers (`DropBlock`, `DropPath`). Currently, according to [Lux's documentation](https://lux.csail.mit.edu/dev/api/layers/#Lux.Chain), there is an option to enable and disable optimisations that can remove these and make the model a little cleaner to go through. Is there a chance something similar can be implemented for Flux?

**Comment by theabhirath** [comment]
Speaking of Lux's documentation, it looks absolutely beautiful - any chance Flux would consider using a different theme?

**Comment by darsnack** [comment]
We could transition to Pollen

**Comment by darsnack** [comment]
As far as optimization goes, I don't think the Lux optimizations do what you're proposing. Instead, they recursively go through the `Chain` to wrap functions that don't adhere to the Lux interface and to delete `NoOpLayer`s. Both of which aren't issues for Flux models.

That being said, it wouldn't be too difficult to build the kind of optimization you're talking about using `fmap` (we wouldn't want the keyword interface that's limited to `Chain` like Lux). The main question is what kind of benefit it provides. If most of these no-ops can be optimized by the compiler itself, then this kind of optimization pass isn't super useful. Though I wouldn't be surprised if it gave a benefit on the backwards pass due to Zygote. Maybe a comparison of a manually optimized vs. un-optimized model from Metalhead would be good to have first.

**Comment by avik-pal** [comment]
Just to answer a few points raised here:
1. The optimization pass is necessary for Lux since it requires layers to follow a particular interface. For Flux, it makes little sense since it doesn't require a strict interface.
2. No-ops would ideally be optimized away. But Zygote being zygote keeps them around. Becomes worse for Dropout where it keeps branching around :disappointed:. Though in most real-world use cases, it makes very little difference. It shows up if your model is reasonably small.

**Comment by ToucheSir** [comment]
> 2\. But Zygote being zygote keeps them around. Becomes worse for Dropout where it keeps branching around

On this point, see https://github.com/FluxML/Flux.jl/pull/2005. I still think making `active` a type parameter as you did is cleaner, but we don't have that luxury here.

**Comment by DhairyaLGandhi** [comment]
Not really. The short templated chains means we can often use the compiled
gradient code across a model often

**Comment by CarloLucibello** [comment]
Closing — as discussed in-thread, Lux`s optimisation pass exists because Lux requires layers to satisfy a strict interface; Flux does not, so there is no equivalent work for it to do. Removing no-op layers is a cosmetic change that does not buy a measurable speedup, and users who want it can filter their own `Chain` before constructing it.
